manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,451 @@
+// being composed of 2 numeric types, this type is comparable and can be used
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
+	port     uint32
+}
+
+// Proposal is the return value of Allocate and DeallocateEndpoint,
+// which contains the data needed to affirmatively alter the state of the
+// Allocator. It exists so that Allocator changes can be abandoned
+// without complex rollback logic. To commit changes to the Allocator, call
+// the proposal's Commit function
+type Proposal interface {
+	// Ports returns the port assignments generated by this proposal
+	Ports() []*api.PortConfig
+
+	// Commit commits the result of the Allocator transaction and alters
+	// the state of the port allocator. Part of the contract of the
+	// Allocator is that it should always return a valid proposal, and so
+	// Commit will always succeed and does not need to return an error.
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify the net
+	// allocator state. IsNoop does not mean the ports for an endpoint haven't
+	// changed. It only means that the publish ports marked in use by the
+	// allocator haven't changed.
+	//
+	// Used mostly for testing, and should probably not be relied on outside of
+	// this package.
+	IsNoop() bool
+}
+
+type proposal struct {
+	pa         *Allocator
+	ports      []*api.PortConfig
+	allocate   map[port]struct{}
+	deallocate map[port]struct{}
+}
+
+func (prop *proposal) Ports() []*api.PortConfig {
+	if prop.ports == nil {
+		return []*api.PortConfig{}
+	}
+	return prop.ports
+}
+
+// Commit commits the proposal to the port allocator.
+func (prop *proposal) Commit() {
+	if prop.IsNoop() {
+		// nothing to do if proposal is noop, short circuit a bit
+		return
+	}
+	// The pattern here is we're going to free every port in p.deallocate and
+	// then allocate every port in allocate. any over lap results in no net
+	// change
+	for p := range prop.deallocate {
+		delete(prop.pa.ports, p)
+	}
+	for p := range prop.allocate {
+		prop.pa.ports[p] = struct{}{}
+	}
+}
+
+// IsNoop returns true if the ports in use before this proposal are the same as
+// the ports in use after.
+func (prop *proposal) IsNoop() bool {
+	// if the allocate and deallocate sets are the same length, and every port
+	// in allocate is also found in deallocate, then this proposal is a noop
+	if len(prop.allocate) != len(prop.deallocate) {
+		return false
+	}
+	for p := range prop.allocate {
+		if _, ok := prop.deallocate[p]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// NewAllocator returns a new instance of the Allocator object
+func NewAllocator() *Allocator {
+	return &Allocator{
+		ports: make(map[port]struct{}),
+	}
+}
+
+// Restore adds the current endpoints to the local state of the port allocator
+// but does not perform any new allocation.
+func (pa *Allocator) Restore(endpoints []*api.Endpoint) {
```

Comment:
Am I correct thinking that this method is only used just after creating the allocator? Taking it to a normal function instead of a "method", creating the allocator inside and returning the allocator looks nicer in my head, as it will ensure that it only gets called once on an empty allocator.

Reply:
there are a few good reasons not to do it like this:

1. the constructor never returns an error, which means that anything that relies on it can in turn avoid returning errors from its constructor, all the way to the top. this is handy because when everything is said and done, it means you can have the top-level allocator component initialize the subcomponents in its constructor, and then run the Restore method in its Run function. it's hard to explain

2. you can call Restore multiple times, which isn't always necessary, but which is useful for testing because subtests can call Restore again with more state.